[KeyVault] adding msrestazure credential authentication support

schaabs opened the pull request:
Updating key vault client to support both KeyVaultAuthentication and msrestazure Credential object for authentication.

lmazuel commented on azure-keyvault/azure/keyvault/custom/key_vault_client.py:
```diff
@@ -6,12 +6,43 @@
+        self._inner_creds = None
+
+        # if the supplied credentials instance is not derived from KeyVaultAuthBase but is an AADTokenCredentials instance
+        if not isinstance(credentials, KeyVaultAuthBase) and isinstance(credentials, AADTokenCredentials):
```
`AADMixin`, not `AADTokenCredentials`. This one is deprecated and not in the hierarchy of `ServicePrincipalCredentials`

lmazuel commented on azure-keyvault/azure/keyvault/custom/key_vault_client.py:
```diff
@@ -6,12 +6,43 @@
+
+            def auth_callback(server, resource, scope):
+                self._inner_creds.resource = resource
+                token = self._inner_creds.token
```
`token` is just a property and does not call AD to get the token, so you don't do the call with the new resource.
Add this before to do the call :
```python
self._inner_creds.set_token()
```

schaabs commented:
@lmazuel I've updated the code based off your feedback and our conversation.   I believe this scheme will work better as it allows KeyVaultAuthentication to handle refresh_session.  I'm still doing some testing but I've updated the PR so you can take a look.

lmazuel commented on azure-keyvault/azure/keyvault/custom/key_vault_authentication.py:
```diff
@@ -140,18 +140,51 @@ def auth_callack(server, resource, scope):
+        :rtype: requests.Session.
+        """
+        if self._credentials:
+            self._credentials.refresh_session()
```
miss a return

RandalliLama commented on azure-keyvault/azure/keyvault/custom/key_vault_authentication.py:
```diff
@@ -140,18 +141,47 @@ def auth_callack(server, resource, scope):
         """
-        super(KeyVaultAuthentication, self).__init__()
+        if not authorization_callback and not credentials:
+            raise ValueError("Either parameter 'authorization_callback' or parameter 'credentials' must not be None.")
```
This seems like a double negative.  How about:

Either parameter 'authorization_callback' or parameter 'credentials' must be specified.

RandalliLama commented on azure-keyvault/azure/keyvault/custom/key_vault_authentication.py:
```diff
@@ -140,18 +141,47 @@ def auth_callack(server, resource, scope):
+                if self._credentials.resource != resource:
+                    self._credentials.resource = resource
+                    token = self._credentials.set_token()
+                token = self._credentials.token
```
I don't understand 165 and 166.  Won't 166 just undo what was done at 165, or is there some Python magic going on here?
